File: core/src/pl/shockah/godwit/gesture/Touch.java
package pl.shockah.godwit.gesture;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import lombok.Getter;
import pl.shockah.godwit.Godwit;
import pl.shockah.godwit.geom.IVec2;
import pl.shockah.godwit.geom.Vec2;

public final class Touch {
	public final int pointer;
	@Nonnull public final List<Point> points = new ArrayList<>();

	@Getter
	@Nullable private ContinuousGestureRecognizer continuousRecognizer = null;

	@Getter
	private boolean finished = false;

	public Touch(int pointer) {
		this.pointer = pointer;
	}

	public void addPoint(float x, float y) {
		addPoint(new Vec2(x, y));
	}

	public void addPoint(@Nonnull IVec2 position) {
		if (position instanceof Vec2)
			addPoint((Vec2)position);
		else
			addPoint(position.getImmutableCopy());
	}

	public void addPoint(@Nonnull Vec2 position) {
		if (finished)
			return;
		points.add(new Point(position));
	}

	public void finish() {
		finished = true;
	}

	public void setContinuousRecognizer(@Nullable ContinuousGestureRecognizer recognizer) {
		continuousRecognizer = recognizer;
		if (recognizer != null) {
			for (ContinuousGestureRecognizer continuousRecognizer : new ArrayList<>(Godwit.getInstance().inputManager.gestureManager.currentContinuousRecognizers)) {
				if (continuousRecognizer == recognizer)
					continue;
				continuousRecognizer.onTouchUsedByContinuousRecognizer(this);
			}
		}
	}

	public static final class Point {
		@Nonnull public final Vec2 position;
		@Nonnull public final Date date;

		public Point(@Nonnull Vec2 position) {
			this(position, new Date());
		}

		public Point(@Nonnull Vec2 position, @Nonnull Date date) {
			this.position = position;
			this.date = date;
		}
	}
}

File: core/src/pl/shockah/godwit/node/gesture/Touch.kt
package pl.shockah.godwit.node.gesture

import pl.shockah.godwit.GDelegates
import pl.shockah.godwit.geom.ImmutableVec2
import pl.shockah.godwit.geom.Vec2
import java.util.*

class Touch(
		val pointer: Int,
		val button: Int
) {
	private val mutablePoints = mutableListOf<Point>()

	var recognizer: GestureRecognizer? by GDelegates.changeObservable(null) { _, _: GestureRecognizer?, new: GestureRecognizer? ->
		new?.stage?.activeContinuousGestureRecognizers?.forEach {
			if (it != new)
				it.onTouchUsedByRecognizer(this)
		}
	}

	var finished = false
		private set

	val points: List<Point>
		get() = mutablePoints

	fun finish() {
		finished = true
	}

	fun addPoint(point: Vec2) {
		mutablePoints += Point(point.immutable(), Date())
	}

	operator fun plusAssign(point: Vec2) {
		addPoint(point)
	}

	data class Point(
			val position: ImmutableVec2,
			val date: Date
	)
}